hail/src/main/scala/is/hail/io/tabix/TabixReader.scala

```diff
@@ -0,0 +1,346 @@
+        nOff += idx._1.get(bins(i)).map(_.length).getOrElse(0)
+        i += 1
+      }
+      if (nOff == 0)
+        new Array[TbiPair](0)
+      else {
+        var off = new Array[TbiPair](nOff)
+        nOff = 0
+        i = 0
+        while (i < bins.length) {
+          val c = idx._1.getOrElse(bins(i), null)
+          val len = if (c == null) { 0 } else { c.length }
+          var j = 0
+          while (j < len) {
+            if (TbiOrd.less64(minOff, c(j)._2)) {
+              off(nOff) = TbiPair(c(j)._1, c(j)._2)
+              nOff += 1
+            }
+            j += 1
+          }
+          i += 1
+        }
+        Arrays.sort(off, 0, nOff, null)
+        // resolve contained adjacent blocks
+        var l = 0
+        i = 1
+        while (i < nOff) {
+          if (TbiOrd.less64(off(l)._2, off(i)._2)) {
+            l += 1
+            off(l)._1 = off(i)._1
+            off(l)._2 = off(i)._2
+          }
+          i += 1
+        }
+        nOff = l + 1
+        // resolve overlaps
+        i = 1
+        while (i < nOff) {
+          if (!TbiOrd.less64(off(i - 1)._2, off(i)._1))
+            off(i - 1)._2 = off(i)._1
+          i += 1
+        }
+        // merge blocks
+        i = 1
+        l = 0
+        while (i < nOff) {
+          if ((off(l)._2 >> 16) == (off(i)._1 >> 16))
+            off(l)._2 = off(i)._2
+          else {
+            l += 1
+            off(l)._1 = off(i)._1
+            off(l)._2 = off(i)._2
+          }
+          i += 1
+        }
+        nOff = l + 1
+        val ret = Array.fill[TbiPair](nOff)(null)
+        i = 0
+        while (i < nOff) {
+          if (off(i) != null)
+            ret(i) = TbiPair(off(i)._1, off(i)._2)
+          i += 1
+        }
+        if (ret.length == 0 || (ret.length == 1 && ret(0) == null))
+          new Array[TbiPair](0)
+        else
+          ret
+      }
+    }
+  }
+
+  private def reg2bins(beg: Int, _end: Int): Array[Int] = {
+    if (beg >= _end)
+      new Array[Int](0)
+    else {
+      var end = _end
+      val bins = new ArrayBuilder[Int](MaxBin)
+      if (end >= (1 << 29)) {
+        end = 1 << 29
+      }
+      end -= 1
+      bins += 0
+      var k = 1 + (beg >> 26)
+      while (k <= 1 + (end >> 26)) {
+        bins += k
+        k += 1
+      }
+      k = 9 + (beg >> 23)
+      while (k <= 9 + (end >> 23)) {
+        bins += k
+        k += 1
+      }
+      k = 73 + (beg >> 20)
+      while (k <= 73 + (end >> 20)) {
+        bins += k
+        k += 1
+      }
+      k = 585 + (beg >> 17)
+      while (k <= 585 + (end >> 17)) {
+        bins += k
+        k += 1
+      }
+      k = 4681 + (beg >> 14)
+      while (k <= 4681 + (end >> 14)) {
+        bins += k
+        k += 1
+      }
+      bins.result()
+    }
+  }
+}
+
+class TabixLineIterator(private val filePath: String, private val offsets: Array[TbiPair])
+  extends java.lang.AutoCloseable
+{
+  private val hConf = HailContext.get.sc.hadoopConfiguration
```

Comment:
is this class meant to be instantiated only on the driver? if so, this is fine. If not, we'll need to take a serialized hadoop configuration in the constructor.